Move PICO to EISOP

aosen-xiong opened the pull request:
Original code repo: https://github.com/opprop/immutability
Some of commits are presented in: https://github.com/Ao-senXiong/immutability/tree/pico-cf-only
Merge with: https://github.com/eisop/jdk/pull/106
As discussed before, I will try to move PICO here gradually. Hopefully adding test cases and Javadoc with more discussion as well.

aosen-xiong committed:
Don't recursively call the function if the underlining type is raw

aosen-xiong committed:
Compute Varargs before calling setParameterTypes

aosen-xiong committed:
Compute Varargs before calling setParameterTypes

aosen-xiong committed:
Starting move PICO to EISOP

Co-authored-by: Werner Dietl <wdietl@gmail.com>
Co-authored-by: Haifeng Shi <shihaifeng1998@gmail.com>
Co-authored-by: Weitian Xing <xingweitian@gmail.com>
Co-authored-by: Jeff Luo <j36luo@uwaterloo.ca>
Co-authored-by: Mier Ta <m2ta@uwaterloo.ca>

aosen-xiong committed:
Remove static import and holder class

aosen-xiong committed:
Clear star import and static import again

aosen-xiong committed:
Lots of Javadoc and remove some unused method

aosen-xiong committed:
Remove dangling Javadoc

aosen-xiong committed:
Add Javadoc

aosen-xiong committed:
A few more Javadoc

aosen-xiong committed:
Add PICO's annotation for format

aosen-xiong committed:
Add PICO's annotation for format

aosen-xiong committed:
Add PICO's annotation for format

aosen-xiong committed:
Fix a few errors and add the very first test cases

aosen-xiong committed:
Fix a few errors and add the very first test cases

aosen-xiong committed:
Format poly and forbid RDM on static class

aosen-xiong committed:
Format poly and forbid RDM on static class

aosen-xiong committed:
Format poly and forbid RDM on static class

aosen-xiong committed:
Yet another test case

aosen-xiong committed:
Yet another test case

aosen-xiong committed:
Add object creation test

aosen-xiong committed:
Add object creation test

aosen-xiong committed:
Add assignability annotation test and improve the error message

aosen-xiong committed:
Add assignability annotation test and improve the error message

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitTransfer.java:
```diff
@@ -0,0 +1,49 @@
+            VariableElement varElement =
+                    TreeUtils.elementFromDeclaration((VariableTree) n.getTarget().getTree());
+            // Below is for removing false positive warning of bottom illegal write cacused by
+            // refining field to @Bottom if
+            // field initializer is null.
```
```suggestion
            // refining field to @Bottom if field initializer is null.
```

wmdietl replied:
Please re-flow the comments.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitVisitor.java:
```diff
@@ -0,0 +1,588 @@
+        }
+    }
+
+    // Disables method overriding checks in BaseTypeVisitor
```
Why can we disable these?

Ao-senXiong replied:
Updated with comment. The implementation is hacky right now as it invoke viewpoint adapter in visitor.

Ao-senXiong replied:
We move the viewpoint adaptation logic to AnnotatedType factory.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/jdk.astub:
```diff
@@ -0,0 +1,355 @@
+import org.checkerframework.checker.pico.qual.Mutable;
```
Now that we're moving PICO into EISOP, we should add these annotations directly into eisop/jdk.
If you use the same branch name, hopefully CI picks up the right branch.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/jdk.astub:
```diff
@@ -0,0 +1,355 @@
+
+@ReceiverDependentMutable
+class Iterable<E> {
+  Iterator<E> iterator(@Readonly Iterable this);
```
```suggestion
  @PolyMutable Iterator<E> iterator(@PolyMutable Iterable this);
```

Should this use `PolyMutable`? We should make sure that for a readonly/immutable Iterable, we cannot produce a mutable Iterator, which the current signature incorrectly allows.
By using PolyMutable, we connect the mutability of the two.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/jdk.astub:
```diff
@@ -0,0 +1,355 @@
+}
+
+class RuntimeException {
+  RuntimeException(String var0, Throwable var1);
```
Should there be an annotation, e.g. on var1?

Ao-senXiong replied:
Added readonly.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/jdk.astub:
```diff
@@ -0,0 +1,355 @@
+  StringBuilder append(@Readonly Object var1);
+}
+
+class StringBuffer {
```
What default is currently used in stub files? If it's also immutable, many of these annotations are wrong.
In general, we should move to having two opt-in annotations, for a mutable or an immutable default, ASAP.

Ao-senXiong replied:
Currently it is immutable. I have updated my annotations in the linked JDK PR.

wmdietl commented on checker/src/main/java/org/checkerframework/checker/pico/jdk.astub:
```diff
@@ -0,0 +1,355 @@
+
+@ReceiverDependentMutable
+interface CharSequence {
+  int length(@Readonly CharSequence this);
```
I wonder whether the default for `CharSequence` should be readonly... the interface itself must be RDM, e.g. to allow StringBuilder, which is mutable. But CharSequence itself is readonly?
So adding [DefaultQualifierForUse](checker-qual/src/main/java/org/checkerframework/framework/qual/DefaultQualifierForUse.java) to such classes might make their use easier?

Ao-senXiong replied:
Done.

wmdietl reviewed (changes requested):
A quick round of initial comments.

aosen-xiong committed:
Use `PolyMutable` for iterator

Co-authored-by: Werner Dietl <wdietl@gmail.com>

aosen-xiong committed:
Use `PolyMutable` for iterator

Co-authored-by: Werner Dietl <wdietl@gmail.com>

Ao-senXiong commented on checker-qual/src/main/java/org/checkerframework/checker/pico/qual/Bottom.java:
```diff
@@ -0,0 +1,29 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE_PARAMETER})
+@TargetLocations({TypeUseLocation.LOWER_BOUND})
+public @interface Bottom {}
```
Thanks, I will use `PICOBottom` and `PICOLost`

Ao-senXiong commented on checker-qual/src/main/java/org/checkerframework/checker/pico/qual/Lost.java:
```diff
@@ -0,0 +1,16 @@
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.TYPE_USE, ElementType.TYPE_PARAMETER})
+public @interface Lost {}
```
`PICOLost` sounds good to me.

Ao-senXiong commented on checker/src/main/java/org/checkerframework/checker/pico/PICOInitializationAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,104 @@
+        return uninitializedFields;
+    }
+
+    // TODO this is a hack for calling static class method for lambda expression, consider remove
```
Previously there is accessibilty issue, I will remove that now.

aosen-xiong committed:
Rename qualifier to `PICO...`

aosen-xiong committed:
Rename qualifier to `PICO...`

aosen-xiong committed:
Remove hack

aosen-xiong committed:
Remove hack

Ao-senXiong commented on checker/src/test/java/org/checkerframework/checker/test/junit/PICOImmutabilityTest.java:
```diff
@@ -0,0 +1,24 @@
+        super(testFiles, org.checkerframework.checker.pico.PICOChecker.class, "pico-immutability");
+    }
+
+    @Parameterized.Parameters
```
That is odd. Before today, it was https://dev.azure.com/eisop/checker-framework/_build/results?buildId=6325&view=logs&jobId=4b19ccca-7d7d-5749-26de-b87349fba2b4&j=4b19ccca-7d7d-5749-26de-b87349fba2b4&t=4dba1f6c-8e7b-5172-04db-4dd554c55b12.

aosen-xiong committed:
Fixes misc-21 error

aosen-xiong committed:
Fixes misc-21 error

Ao-senXiong commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,749 @@
+ * TypeAnnotators. It also applies implicits to method receiver that is not so by default in super
+ * implementation.
+ */
+// TODO Use @Immutable for classes that extends those predefined immutable classess like String or
```
Okay, make sense to me.

aosen-xiong committed:
Also test validity of annotation on interface and abstract class

aosen-xiong committed:
Also test validity of annotation on interface and abstract class

aosen-xiong committed:
Remove wrong todo and add JML readonly aliasing

aosen-xiong committed:
Remove wrong todo and add JML readonly aliasing

Ao-senXiong commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitAnnotatedTypeFactory.java:
```diff
@@ -0,0 +1,749 @@
+        super(checker);
+        postInit();
+        // PICO aliasing is not implemented correctly remove for now
+        // addAliasedAnnotation(org.jmlspecs.annotation.Readonly.class, READONLY);
```
Thanks!

aosen-xiong committed:
Apply spotless

aosen-xiong committed:
Apply spotless

Ao-senXiong commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitVisitor.java:
```diff
@@ -0,0 +1,588 @@
+        PICOViewpointAdapter vpa = atypeFactory.getViewpointAdapter();
+        AnnotationMirror adapted = vpa.combineAnnotationWithAnnotation(lhs, rhs);
+        return atypeFactory.getQualifierHierarchy().isSubtypeQualifiersOnly(adapted, lhs);
+    }
```
After #1032, this method and the override of `isValidUse(
            AnnotatedDeclaredType declarationType, AnnotatedDeclaredType useType, Tree tree)` can be removed.

aosen-xiong replied:
Resolved based on #1032.

We cannot remove the `isValidUse` override entirely: `BaseTypeVisitor` checks `use <: (use |> bound(C))`, while mutability needs the well-formedness rule `(q |> bound(C)) <: q` so that `@Readonly` (and similar) uses of mutable / receiver-dependent classes remain valid.

What we can remove is the hand-rolled `isAdaptedSubtype` helper. The override now uses `AnnotatedTypeFactory#getTypeDeclarationBoundsFromUse` from #1032 and checks the adapted bounds with `isSubtypeShallowEffective(adaptedBounds, useType)`.

aosen-xiong committed:
Add RDM class test and initialization test

aosen-xiong committed:
Add RDM class test and initialization test

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Add implicit mutable test

aosen-xiong committed:
Add implicit mutable test

aosen-xiong committed:
Add Definite Assignment test

aosen-xiong committed:
Add Definite Assignment test

aosen-xiong committed:
Delete Definite Assignment test as it is already tested in `PICOInitialization`

aosen-xiong committed:
Delete Definite Assignment test as it is already tested in `PICOInitialization`

aosen-xiong committed:
Add test for type use and inheritance

aosen-xiong committed:
Add test for type use and inheritance

aosen-xiong committed:
A few small improvement

aosen-xiong committed:
A few small improvement

aosen-xiong committed:
Add lhs to test object creation default

aosen-xiong committed:
Add lhs to test object creation default

aosen-xiong committed:
Add builder pattern test case

aosen-xiong committed:
Add builder pattern test case

aosen-xiong committed:
Additional error message

aosen-xiong committed:
Additional error message

aosen-xiong committed:
Add viewpoint adaptation test

aosen-xiong committed:
Add viewpoint adaptation test

aosen-xiong committed:
Add operator test

aosen-xiong committed:
Add operator test

aosen-xiong committed:
Add immutableclass and immutable list test cases

aosen-xiong committed:
Add immutableclass and immutable list test cases

aosen-xiong committed:
Add cast, enum and refine the RDM test cases

aosen-xiong committed:
Add cast, enum and refine the RDM test cases

aosen-xiong committed:
Move some more operator test here

aosen-xiong committed:
Move some more operator test here

aosen-xiong committed:
Add adapt constructor parameter test, it is adapted correctly but method parameters are not

aosen-xiong committed:
Add adapt constructor parameter test, it is adapted correctly but method parameters are not

aosen-xiong committed:
Move all test cases over first

aosen-xiong committed:
Move all test cases over first

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Remove committed wrong file

aosen-xiong committed:
Remove committed wrong file

Ao-senXiong commented on checker/tests/pico-immutability/ViewpointAdaptationRules.java:
```diff
@@ -0,0 +1,178 @@
+
+        this.assignableReadonlyField = rdmObject;
+        // :: error: (assignment.type.incompatible) :: error: (illegal.field.write)
+        this.rdmField = rdmObject; // Field is adpated to immutable but method parameters are not?
```
This is wrong, method parameter should also be adapted. #1037

Ao-senXiong replied:
visitMethod is getting its oracle in `AnnotatedExecutableType methodType = atypeFactory.getAnnotatedType(tree).deepCopy();`. It should get adapated version of the original method signature.

aosen-xiong committed:
Rerun the test as it should use JDK from same branch name

aosen-xiong committed:
Rerun the test as it should use JDK from same branch name

Ao-senXiong commented:
Ah, looks like it can not find the branch again.
```
Cloning into '.plume-scripts'...
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'pico-move': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
About to run: git clone -b master -q --single-branch --depth 1 https://github.com/eisop/jdk.git ../jdk
git-clone-related: ../jdk is at 2a5380ebe75fbfc4a26169a682f0418694aa6d50
```

Ao-senXiong commented:
Looks like the CI issue not only happens for this PR, also for #1034. It did not fail because it does not have other related PR and fallback to master branches works fine.

```
# Clone the annotated JDK into ../jdk .
"$GIT_SCRIPTS/git-clone-related" ${DEBUG_FLAG} eisop jdk
++ /home/runner/work/checker-framework/checker-framework/checker/bin-devel/.git-scripts/git-clone-related eisop jdk
In directory /home/runner/work/checker-framework/checker-framework :
Cloning into '.plume-scripts'...
fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
fatal: ambiguous argument 'eisop-issue-[103](https://github.com/eisop/checker-framework/pull/1034/checks#step:6:104)3': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
About to run: git clone -b master -q --single-branch --depth 1 https://github.com/eisop/jdk.git ../jdk
```

aosen-xiong committed:
Test with CI Debug

aosen-xiong committed:
Test with CI Debug

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Fix errors for JDK 11

aosen-xiong committed:
Fix errors for JDK 11

Ao-senXiong commented:
I think we have to bear the windows failure as the script does not support windows
```
++ /d/a/checker-framework/checker-framework/checker/bin-devel/.git-scripts/git-find-fork Ao-senXiong eisop jdk
+ '[' 3 -ne 3 ']'
+ ORG=Ao-senXiong
+ UPSTREAM_ORG=eisop
+ REPONAME=jdk
+ export 'GITEXISTS=wget -q --spider'
+ GITEXISTS='wget -q --spider'
+ wget -q --spider https://github.com/Ao-senXiong/jdk.git
/d/a/checker-framework/checker-framework/checker/bin-devel/.git-scripts/git-find-fork: line 29: wget: command not found
+ OWNER=eisop
+ echo https://github.com/eisop/jdk.git
+ REPO_URL=https://github.com/eisop/jdk.git
+ '[' -n 1 ']'
git-find-fork Ao-senXiong eisop jdk => https://github.com/eisop/jdk.git
+ echo 'git-find-fork Ao-senXiong eisop jdk => https://github.com/eisop/jdk.git'
+ '[' -n 1 ']'
+ echo 'About to run git-find-branch' https://github.com/eisop/jdk.git pico-move
About to run git-find-branch https://github.com/eisop/jdk.git pico-move
++ /d/a/checker-framework/checker-framework/checker/bin-devel/.git-scripts/git-find-branch https://github.com/eisop/jdk.git pico-move
+ REPO_URL=https://github.com/eisop/jdk.git
+ shift
+ for BRANCH in "$@"
+ git ls-remote --quiet --exit-code --heads https://github.com/eisop/jdk.git pico-move
+ git ls-remote --symref https://github.com/eisop/jdk.git HEAD
+ awk '/^ref:/ {sub(/refs\/heads\//, "", $2); print $2}'
+ REPO_BRANCH=master
+ '[' -n 1 ']'
+ echo 'git-find-branch https://github.com/eisop/jdk.git pico-move => master'
+ '[' https://github.com/eisop/jdk.git '!=' https://github.com/eisop/jdk.git ']'
git-find-branch https://github.com/eisop/jdk.git pico-move => master
+ echo 'About to run: git clone -b master -q --single-branch --depth 1 https://github.com/eisop/jdk.git ../jdk'
+ timeout 60 git clone --depth 1 -b master -q --single-branch --depth 1 https://github.com/eisop/jdk.git ../jdk
About to run: git clone -b master -q --single-branch --depth 1 https://github.com/eisop/jdk.git ../jdk```

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Remove unused override

aosen-xiong committed:
Remove unused override

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Correct the test cases

aosen-xiong committed:
Correct the test cases

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Undo debug

aosen-xiong committed:
Undo debug

aosen-xiong committed:
Test array as field in class

aosen-xiong committed:
Test array as field in class

aosen-xiong committed:
Delete jdk.astub

aosen-xiong committed:
Delete jdk.astub

aosen-xiong committed:
Correct code and refine example

aosen-xiong committed:
Correct code and refine example

aosen-xiong committed:
Partially fixed override rule

aosen-xiong committed:
Partially fixed override rule

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Remove hack for adding annotation for toString and other methods

aosen-xiong committed:
Remove hack for adding annotation for toString and other methods

Ao-senXiong commented on checker/tests/pico-immutability/ClassAnnotation.java:
```diff
@@ -0,0 +1,232 @@
+
+    // TODO Make the superclass's bound implicit works
+    // :: error: (declaration.inconsistent.with.extends.clause) :: error: (super.invocation.invalid)
+    class ImmutableChildClassBad1 extends @Mutable MutableClass {}
```
Forbid annotation before extends/implements clause.

aosen-xiong committed:
Apply suggestions from code review

Co-authored-by: Werner Dietl <wdietl@gmail.com>

aosen-xiong committed:
Apply suggestions from code review

Co-authored-by: Werner Dietl <wdietl@gmail.com>

aosen-xiong committed:
Remove super clause annotator

aosen-xiong committed:
Remove super clause annotator

aosen-xiong committed:
Remove used code

aosen-xiong committed:
Remove used code

aosen-xiong committed:
Apply spotless

aosen-xiong committed:
Apply spotless

Ao-senXiong commented:
Looks like this hack override of `checkoverride` I removed in [38a547a](https://github.com/eisop/checker-framework/pull/1013/commits/38a547a78a990bdfa6f1d6b00b3e5e75e718a9ce) also indicates a bug in the framework `StructuralEqualityComparer`. Upstream has similar 6867.

aosen-xiong committed:
Fixes PolyMutable factory pattern

aosen-xiong committed:
Fixes PolyMutable factory pattern

aosen-xiong committed:
Merge branch 'master' into pico-move

Ao-senXiong commented on checker/src/main/java/org/checkerframework/checker/pico/PICONoInitVisitor.java:
```diff
@@ -0,0 +1,557 @@
+    }
+
+    @Override
+    protected boolean isTypeCastSafe(AnnotatedTypeMirror castType, AnnotatedTypeMirror exprType) {
+        QualifierHierarchy qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+
+        final TypeKind castTypeKind = castType.getKind();
+        if (castTypeKind == TypeKind.DECLARED) {
+            // Don't issue an error if the mutability annotations are equivalent to the qualifier
+            // upper bound of the type.
+            // BaseTypeVisitor#isTypeCastSafe is not used, to be consistent with inference which
+            // only have mutability qualifiers if inference is supporting FBC in the future, this
+            // overridden method can be removed.
+            AnnotatedDeclaredType castDeclared = (AnnotatedDeclaredType) castType;
+            AnnotationMirror bound =
+                    qualifierHierarchy.findAnnotationInHierarchy(
+                            atypeFactory.getTypeDeclarationBounds(castDeclared.getUnderlyingType()),
+                            atypeFactory.READONLY);
+            assert bound != null;
+
+            if (AnnotationUtils.areSame(
+                    castDeclared.getAnnotationInHierarchy(atypeFactory.READONLY), bound)) {
+                return true;
+            }
+        }
+
+        return super.isTypeCastSafe(castType, exprType);
```
This should be nicer after #1049

aosen-xiong committed:
Commit the hack override

aosen-xiong committed:
Doc cleanup

aosen-xiong committed:
Doc cleanup

aosen-xiong committed:
Clean up duplicate code for constructor invocation

aosen-xiong committed:
Clean up duplicate code for constructor invocation

aosen-xiong committed:
Fixed crash bugs and a few more cleanups

aosen-xiong committed:
Fixed crash bugs and a few more cleanups

aosen-xiong committed:
Clean up code and default method receiver to its class bound

aosen-xiong committed:
Clean up code and default method receiver to its class bound

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
PR 1078

aosen-xiong committed:
Fix behavior in transitive mutable field

aosen-xiong committed:
Fix behavior in transitive mutable field

aosen-xiong committed:
Fixes array component type

aosen-xiong committed:
Fixes array component type

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Adapt framework change

aosen-xiong committed:
Adapt framework change

aosen-xiong committed:
Fixing newest change

aosen-xiong committed:
Fixing newest change

aosen-xiong committed:
Allow underinit receiver to do assignment

aosen-xiong committed:
Allow underinit receiver to do assignment

aosen-xiong committed:
Add test case

aosen-xiong committed:
Add test case

aosen-xiong committed:
Add Circular initialization test case

aosen-xiong committed:
Add Circular initialization test case

aosen-xiong committed:
Allow static class be annotated as RDM

aosen-xiong committed:
Allow static class be annotated as RDM

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Add `assumeInitialized` option

aosen-xiong committed:
Add `assumeInitialized` option

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Add suppress warning string

aosen-xiong committed:
Add suppress warning string

aosen-xiong committed:
Lost adapts RDM

aosen-xiong committed:
Init change

aosen-xiong committed:
Get default value for poly qualifier

aosen-xiong committed:
Use fully qualified name for field

aosen-xiong committed:
Expand test cases

aosen-xiong committed:
Further expand the test cases with generics

aosen-xiong committed:
Add comment for new field in polynull

aosen-xiong committed:
Undo changes not needed

aosen-xiong committed:
PolyMutable with element

aosen-xiong committed:
Use areSameByName for polymutable in VPA

aosen-xiong committed:
Update default

aosen-xiong committed:
Delete temp files; Spotless

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Don't comit the log

aosen-xiong committed:
Delete class files

aosen-xiong committed:
Don't check initialization for primitive types

aosen-xiong committed:
Some refactor

aosen-xiong committed:
Mutable default tests

aosen-xiong committed:
Add suppress warning for clone

aosen-xiong committed:
Add suppress warning for poly interacts with varargs

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Delete files not needed and format files

aosen-xiong committed:
Merge branch 'pico-move' into pico-move-hack

aosen-xiong committed:
Fixes error prone warnings and format test cases

aosen-xiong committed:
Suppress escape warning

aosen-xiong committed:
Don't run immutable default tests

aosen-xiong committed:
Delete files not needed

aosen-xiong committed:
Fixes eisop#433

aosen-xiong committed:
Remove typetool link

aosen-xiong committed:
Don't create top

aosen-xiong committed:
Add warning

aosen-xiong committed:
Also fix for recursive bound

aosen-xiong committed:
Unmerged PR #1438 to fix crash

aosen-xiong committed:
LazyFinal

aosen-xiong committed:
Add Glacier test

aosen-xiong committed:
Remove OverrideEquals error

aosen-xiong committed:
Default exception parameter to mutable

aosen-xiong committed:
WIP

aosen-xiong committed:
Cleanup LazyFinal

aosen-xiong committed:
Constrictor benchmark

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Address PICO review cleanup

aosen-xiong committed:
Reject receiver-dependent mutable in static contexts

aosen-xiong committed:
Clean up PICO mutability checks

aosen-xiong committed:
Use robust anonymous class detection in PICO

aosen-xiong committed:
Merge pico-artifact into pico-move

aosen-xiong committed:
Revert raw type propagation workaround

aosen-xiong committed:
Restore nullness jtreg class files

aosen-xiong committed:
Address PICO artifact review comments

aosen-xiong committed:
Revert qualifier polymorphism changes

aosen-xiong committed:
Remove local notes from PICO artifact tests

aosen-xiong committed:
Remove PICO object identity helpers

aosen-xiong committed:
Fix PICO flexible override checking

aosen-xiong committed:
Reject static receiver-dependent mutable arrays

aosen-xiong committed:
Remove old PICO immutability test suites

aosen-xiong committed:
Refine PICO implementation and manual

aosen-xiong committed:
Add PICO getClass flow refinement

aosen-xiong committed:
Temporary diagnostic for related JDK clone

aosen-xiong committed:
Test patched git-scripts for related clone

aosen-xiong committed:
Revert temporary related clone diagnostics

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Rename PICO checker to Mutability checker

aosen-xiong committed:
Trigger CI

aosen-xiong committed:
Rename mutability manual figure

aosen-xiong committed:
Report MutabilityLost with targeted errors

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Use getTypeDeclarationBoundsFromUse in mutability isValidUse

After #1032, drop the local isAdaptedSubtype helper and apply the
well-formedness rule (q |> bound) <: q via the shared API. Keep the
PolyMutable/Lost early return; remove the outdated FIXME.

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Viewpoint-adapt overridden method types

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Restrict Mutability initialization writes

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Clean up mutability initialization handling

aosen-xiong committed:
Adapt field initializer LHS types to enclosing viewpoint

aosen-xiong committed:
Separate field initializer viewpoint test

aosen-xiong committed:
Simplify field initializer viewpoint adaptation

aosen-xiong committed:
Test generic field initializer viewpoint adaptation

aosen-xiong committed:
Refine Mutability Checker validation

aosen-xiong committed:
Remove redundant Mutability factory logic

aosen-xiong committed:
Document Mutability supertype adaptation

aosen-xiong committed:
Simplify Mutability defaulting

aosen-xiong committed:
Fix Mutability defaulting paths

aosen-xiong committed:
Recognize synthetic array clone methods

aosen-xiong committed:
Update Mutability benchmark suppressions

aosen-xiong committed:
Avoid Java 9 collection factory

aosen-xiong committed:
Use an immutable set for unary operators

aosen-xiong commented on framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java:
```diff
@@ -1996,6 +1996,20 @@ public AnnotatedTypeMirror getAnnotatedTypeLhs(Tree lhsTree) {
         }
+
+        // A field declaration initializer assigns to an implicit this.field. Viewpoint-adapt the
+        // field's declared type to the enclosing class just as for an explicit member access.
```
clean up after https://github.com/eisop/checker-framework/pull/1900.

aosen-xiong commented on framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java:
```diff
@@ -4674,6 +4674,11 @@ protected boolean checkOverride(
         }
 
+        // Viewpoint-adapt the overridden method type to the overriding class.
+        overriddenMethodType =
```
clean up after https://github.com/eisop/checker-framework/pull/1891.

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Use TreeUtils.isAnonymousClass instead of TypesUtils.isAnonymous

Replace TypesUtils.isAnonymous(TreeUtils.typeOf(...)) with
TreeUtils.isAnonymousClass(ClassTree) in MutabilityNoInitVisitor and
MutabilityValidator, following the API unification in eisop/checker-framework#1909
which fixes a JDK 11 anonymous class detection bug.

aosen-xiong committed:
Resolve TODO: allow @Mutable on enum declarations

Java enums can have mutable fields, so @Mutable on enum declarations
is a valid use case. The existing defaulting (MutabilityEnumDefaultAnnotator)
already handles this correctly via addMissingAnnotations. Added a Counter
enum test case demonstrating mutable enum with mutable state.

aosen-xiong committed:
Resolve TODO: document builder pattern limitation

Replace actionable TODO with descriptive comment explaining that the
modern builder pattern (mutating target via setters) correctly reports
illegal.field.write and would require uniqueness tracking to support.

aosen-xiong committed:
Resolve TODOs: poly invocation errors and chained poly calls

No custom poly.invocation.error key needed — the generic
assignment.type.incompatible is correct for poly return constrained
by receiver mutability. Chained poly calls (test5) correctly error
since each poly return resolves based on the receiver's mutability.

aosen-xiong committed:
Resolve TODO: annotate clone() return as @Immutable

In @Immutable ObjectMethods2, annotate clone() return type as
@Immutable Object instead of using SuppressWarnings workaround.
The override.return.invalid remains (consistent with all three
test classes) due to how Object.clone() is annotated in stubs.

aosen-xiong committed:
Resolve TODO: document varargs limitation

Replace the TODO with a descriptive comment explaining that varargs
type inference with @PolyMutable (e.g. Arrays.asList) does not
correctly adapt to the required mutability context, causing a false
positive that requires @SuppressWarnings.

aosen-xiong committed:
Resolve TODO: document missing ensuresAssigned feature

Convert actionable TODOs into a descriptive note explaining that the
test is skipped because PICO lacks postcondition qualifiers (like
@EnsuresAssigned), preventing it from verifying initialization via
helper methods.

aosen-xiong committed:
Resolve TODO: complete local variable refinement test

Add method calls with assertions to LocalVariableRefinement.java foo()
to test that local variable mutability flow refinement is correctly
applied after reassignment.

aosen-xiong committed:
Resolve TODO: forbid constructors from mutating inherited fields

Update MutabilityNoInitVisitor to strictly check that constructors
can only initialize fields declared in their own class. Reassigning
inherited fields is now forbidden since the superclass constructor
has already completed and its state should not be mutated.
Added illegal.field.write expectation to SuperClassTest.java.

aosen-xiong committed:
Resolve TODO: clarify missing override error

Update MemoizedRectangle.java to replace the TODO with a comment
explaining why no override error is emitted: mutating an @Assignable
field does not require a @Mutable receiver. Also explicitly annotated
the receiver of size() as @Readonly to make it clear why it validly
overrides the superclass method.

aosen-xiong committed:
Fix Java 11 compilation error in test

Make inner classes static in ImplicitMutable.java. In Java 11,
non-static inner classes cannot declare static fields, which was
causing a compilation error in the test suite.

aosen-xiong committed:
Apply spotless

aosen-xiong committed:
Remove obsolete skip-test files ChainOfCallLost and CloneTest

aosen-xiong committed:
Merge branch 'master' into pico-move

aosen-xiong committed:
Fix after #1891
